File: apps/web/src/lib/engine/quantify.ts
// Port of apps/engine/quantify.py -- wasted miles / fleet-hours / dollars.
// Each redundant truck is charged one depot->stop->depot out-and-back.

import type { EngineConfig } from "@/lib/config";
import { distanceMiles } from "@/lib/engine/geo";
import type { CandidateGroup } from "@/lib/engine/detect";
import type { ConsolidatedPlan, RunTotals } from "@/lib/types";

export type QuantifiedGroup = {
  group: CandidateGroup;
  redundant_trucks: number;
  leg_miles: number;
  wasted_miles: number;
  wasted_fleet_hours: number;
  cost_internal: number;
  cost_3pl_benchmark: number;
};

export type QuantifyResult = {
  groups: QuantifiedGroup[];
  totals: Omit<RunTotals, "records_analyzed" | "records_skipped_no_coords">;
};

const round2 = (n: number) => Math.round(n * 100) / 100;

export function quantify(
  groups: CandidateGroup[],
  config: EngineConfig,
): QuantifyResult {
  const c = config.costs;
  const depot: [number, number] = [c.depot.lat, c.depot.lng];
  const serviceHours = c.service_time_minutes / 60;
  const mileRate = c.cost_per_mile + (c.fuel_surcharge_per_mile ?? 0);

  const out: QuantifiedGroup[] = groups.map((g) => {
    // Redundant trucks = trucks used minus the fewest that could legally have
    // carried the combined weight (weight-aware, from detect).
    const redundant = g.distinct_trucks - g.min_trucks_needed;
    const leg = distanceMiles(depot, g.centroid);
    const wastedMiles = redundant * 2 * leg;
    const wastedHours =
      (c.avg_speed_mph ? wastedMiles / c.avg_speed_mph : 0) +
      redundant * serviceHours;
    const costInternal =
      wastedMiles * mileRate + wastedHours * c.cost_per_fleet_hour;
    const cost3pl = wastedMiles * c.third_party_rate_per_mile;

    return {
      group: g,
      redundant_trucks: redundant,
      leg_miles: round2(leg),
      wasted_miles: round2(wastedMiles),
      wasted_fleet_hours: round2(wastedHours),
      cost_internal: round2(costInternal),
      cost_3pl_benchmark: round2(cost3pl),
    };
  });

  const sum = (f: (q: QuantifiedGroup) => number) =>
    round2(out.reduce((a, q) => a + f(q), 0));

  const truckVisitsBefore = out.reduce(
    (a, q) => a + q.group.distinct_trucks,
    0,
  );

  return {
    groups: out,
    totals: {
      candidate_groups: out.length,
      redundant_trucks: out.reduce((a, q) => a + q.redundant_trucks, 0),
      wasted_miles: sum((q) => q.wasted_miles),
      wasted_fleet_hours: sum((q) => q.wasted_fleet_hours),
      cost_internal: sum((q) => q.cost_internal),
      cost_3pl_benchmark: sum((q) => q.cost_3pl_benchmark),
      truck_visits_before: truckVisitsBefore,
      // after = fewest legal trucks per group (weight-aware), not always 1
      truck_visits_after: out.reduce((a, q) => a + q.group.min_trucks_needed, 0),
      truck_visits_eliminated:
        truckVisitsBefore -
        out.reduce((a, q) => a + q.group.min_trucks_needed, 0),
    },
  };
}

// Build the consolidated_plan_json payload for a findings row.
export function toPlan(q: QuantifiedGroup): ConsolidatedPlan {
  return {
    group_id: q.group.group_id,
    type: q.group.type,
    truck_ids: q.group.truck_ids,
    customer_names: q.group.customer_names,
    order_ids: q.group.order_ids.map(String),
    delivery_count: q.group.delivery_count,
    distinct_trucks: q.group.distinct_trucks,
    total_weight_lbs: q.group.total_weight_lbs,
    min_trucks_needed: q.group.min_trucks_needed,
    leg_miles: q.leg_miles,
    centroid: q.group.centroid,
    cost_3pl_benchmark: q.cost_3pl_benchmark,
  };
}

File: apps/engine/quantify.py
"""
quantify.py -- turn candidate groups into wasted miles, fleet-hours, and dollars.

Waste model (Phase 0, deliberately simple and defensible)
---------------------------------------------------------
For a candidate group served by N distinct trucks, one truck was enough. The
other (N-1) trucks each made a separate out-and-back from the depot to reach the
same place. Consolidating removes those redundant out-and-backs.

  redundant_trucks   = distinct_trucks - 1
  leg_miles          = geodesic(depot, group centroid)
  wasted_miles       = redundant_trucks * 2 * leg_miles          # there & back
  wasted_fleet_hours = wasted_miles / avg_speed_mph
                       + redundant_trucks * (service_time_minutes / 60)
  cost_internal      = wasted_miles * cost_per_mile
                       + wasted_fleet_hours * cost_per_fleet_hour
  cost_3pl_benchmark = wasted_miles * third_party_rate_per_mile

`cost_internal` is the primary savings number (what it costs Boise Cascade to run
the redundant trips on their own fleet). `cost_3pl_benchmark` is a reference point:
what those same miles would cost at the 3PL rate.
"""

from __future__ import annotations

from geocode import distance_miles


def quantify(groups: list[dict], config: dict) -> dict:
    """Return {'groups': [...per-group metrics...], 'totals': {...}, 'rates': {...}}."""
    costs = config.get("costs", {})
    depot = costs.get("depot", {})
    depot_pt = (float(depot["lat"]), float(depot["lng"]))

    cost_per_mile = float(costs.get("cost_per_mile", 0.0))
    cost_per_hour = float(costs.get("cost_per_fleet_hour", 0.0))
    tpl_per_mile = float(costs.get("third_party_rate_per_mile", 0.0))
    avg_speed = float(costs.get("avg_speed_mph", 30.0))
    service_hours = float(costs.get("service_time_minutes", 0.0)) / 60.0

    per_group = []
    for g in groups:
        redundant = g["distinct_trucks"] - 1
        leg_miles = distance_miles(depot_pt, g["centroid"])
        wasted_miles = redundant * 2.0 * leg_miles
        wasted_hours = (wasted_miles / avg_speed if avg_speed else 0.0) + redundant * service_hours
        cost_internal = wasted_miles * cost_per_mile + wasted_hours * cost_per_hour
        cost_3pl = wasted_miles * tpl_per_mile

        per_group.append({
            **g,
            "redundant_trucks": redundant,
            "leg_miles": round(leg_miles, 2),
            "wasted_miles": round(wasted_miles, 2),
            "wasted_fleet_hours": round(wasted_hours, 2),
            "cost_internal": round(cost_internal, 2),
            "cost_3pl_benchmark": round(cost_3pl, 2),
        })

    totals = {
        "candidate_groups": len(per_group),
        "redundant_trucks": sum(x["redundant_trucks"] for x in per_group),
        "wasted_miles": round(sum(x["wasted_miles"] for x in per_group), 2),
        "wasted_fleet_hours": round(sum(x["wasted_fleet_hours"] for x in per_group), 2),
        "cost_internal": round(sum(x["cost_internal"] for x in per_group), 2),
        "cost_3pl_benchmark": round(sum(x["cost_3pl_benchmark"] for x in per_group), 2),
        # before/after in truck-visits to the flagged locations
        "truck_visits_before": sum(x["distinct_trucks"] for x in per_group),
        "truck_visits_after": len(per_group),  # one truck per group
    }
    totals["truck_visits_eliminated"] = totals["truck_visits_before"] - totals["truck_visits_after"]

    rates = {
        "cost_per_mile": cost_per_mile,
        "cost_per_fleet_hour": cost_per_hour,
        "third_party_rate_per_mile": tpl_per_mile,
        "avg_speed_mph": avg_speed,
        "service_time_minutes": float(costs.get("service_time_minutes", 0.0)),
        "currency": costs.get("currency", "USD"),
        "depot": depot,
    }

    print(f"[quantify] wasted {totals['wasted_miles']} mi, "
          f"{totals['wasted_fleet_hours']} fleet-hrs, "
          f"${totals['cost_internal']:,.2f} internal "
          f"(${totals['cost_3pl_benchmark']:,.2f} at 3PL rate).")
    return {"groups": per_group, "totals": totals, "rates": rates}
